File: server/src/services/emergencyService.js
/**
 * Emergency Detection Service
 * Config-driven severity classification with escalation rules.
 */

// Severity rules — easy to update without code changes
const SEVERITY_RULES = {
    // Disease name (lowercase) → severity override
    severe: [
        'melanoma', 'melanoma (warning)', 'squamous cell carcinoma',
        'basal cell carcinoma', 'necrotizing fasciitis', 'severe infection',
        'cellulitis', 'sepsis', 'merkel cell carcinoma'
    ],
    moderate: [
        'psoriasis', 'eczema', 'dermatitis', 'rosacea', 'vitiligo',
        'shingles', 'herpes zoster', 'impetigo'
    ],
    mild: [
        'acne', 'ringworm', 'fungal infection', 'tinea', 'warts',
        'contact dermatitis', 'heat rash', 'dry skin'
    ]
};

// Emergency alert thresholds
const EMERGENCY_THRESHOLD = {
    minConfidence: 0.5,          // Only trigger emergency if confidence >= 50%
    severeConditions: SEVERITY_RULES.severe
};

/**
 * Classify severity based on detected condition name.
 * @param {string} conditionName - The detected disease name
 * @param {number} confidenceScore - AI confidence (0-1)
 * @returns {{ severity: string, isEmergency: boolean, alertMessage: string|null }}
 */
function classifySeverity(conditionName, confidenceScore = 0) {
    const normalizedName = (conditionName || '').toLowerCase().trim();
    
    let severity = 'mild'; // default

    // Check severe conditions first
    if (SEVERITY_RULES.severe.some(d => normalizedName.includes(d))) {
        severity = 'severe';
    } else if (SEVERITY_RULES.moderate.some(d => normalizedName.includes(d))) {
        severity = 'moderate';
    } else if (SEVERITY_RULES.mild.some(d => normalizedName.includes(d))) {
        severity = 'mild';
    }

    // Emergency detection
    const isEmergency = severity === 'severe' && confidenceScore >= EMERGENCY_THRESHOLD.minConfidence;

    return {
        severity,
        isEmergency,
        alertMessage: isEmergency
            ? `⚠️ EMERGENCY: Possible serious condition "${conditionName}" detected with ${Math.round(confidenceScore * 100)}% confidence. Please consult a doctor immediately.`
            : null,
        recommendedAction: isEmergency
            ? 'Seek immediate medical consultation. Visit the nearest hospital or call emergency services.'
            : severity === 'moderate'
                ? 'Schedule an appointment with a dermatologist within the next few days.'
                : 'Monitor the condition and maintain good skincare practices.'
    };
}

/**
 * Get the list of severe conditions (for frontend reference).
 */
function getSevereConditions() {
    return SEVERITY_RULES.severe;
}

/**
 * Check if a condition requires emergency alert.
 */
function requiresEmergencyAlert(conditionName, confidenceScore) {
    const result = classifySeverity(conditionName, confidenceScore);
    return result.isEmergency;
}

module.exports = {
    classifySeverity,
    getSevereConditions,
    requiresEmergencyAlert,
    SEVERITY_RULES
};

File: vision-service/src/main/java/com/medivision/service/EmergencyService.java
package com.medivision.service;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.util.*;

/**
 * Emergency Detection Service — config-driven severity classification.
 * Maps detected conditions to severity levels and determines emergency escalation.
 */
@Service
public class EmergencyService {

    private static final Logger log = LoggerFactory.getLogger(EmergencyService.class);

    // Severe conditions that trigger emergency alerts
    private static final Set<String> SEVERE_CONDITIONS = Set.of(
            "melanoma", "melanoma (warning)", "squamous cell carcinoma",
            "basal cell carcinoma", "necrotizing fasciitis", "severe infection",
            "cellulitis", "sepsis", "merkel cell carcinoma"
    );

    // Moderate conditions
    private static final Set<String> MODERATE_CONDITIONS = Set.of(
            "psoriasis", "eczema", "dermatitis", "rosacea", "vitiligo",
            "shingles", "herpes zoster", "impetigo"
    );

    private static final double EMERGENCY_CONFIDENCE_THRESHOLD = 0.5;

    /**
     * Classify the severity of a detected condition.
     *
     * @param conditionName  the detected disease name
     * @param confidenceScore AI confidence (0-1)
     * @return map containing severity, isEmergency, alertMessage, recommendedAction
     */
    public Map<String, Object> classifySeverity(String conditionName, double confidenceScore) {
        String normalized = conditionName != null ? conditionName.toLowerCase().trim() : "";

        String severity = "mild";

        // Check severe conditions first
        for (String severe : SEVERE_CONDITIONS) {
            if (normalized.contains(severe)) {
                severity = "severe";
                break;
            }
        }

        if ("mild".equals(severity)) {
            for (String moderate : MODERATE_CONDITIONS) {
                if (normalized.contains(moderate)) {
                    severity = "moderate";
                    break;
                }
            }
        }

        boolean isEmergency = "severe".equals(severity) && confidenceScore >= EMERGENCY_CONFIDENCE_THRESHOLD;

        Map<String, Object> result = new LinkedHashMap<>();
        result.put("severity", severity);
        result.put("isEmergency", isEmergency);
        result.put("alertMessage", isEmergency
                ? String.format("EMERGENCY: Possible serious condition \"%s\" detected with %.0f%% confidence. Please consult a doctor immediately.",
                conditionName, confidenceScore * 100)
                : null);
        result.put("recommendedAction", isEmergency
                ? "Seek immediate medical consultation. Visit the nearest hospital."
                : "moderate".equals(severity)
                ? "Schedule an appointment with a dermatologist."
                : "Monitor the condition and maintain good skincare.");

        log.info("Severity classification: condition={}, severity={}, emergency={}", conditionName, severity, isEmergency);
        return result;
    }

    public Set<String> getSevereConditions() {
        return SEVERE_CONDITIONS;
    }
}
